crates/core/tedge_mapper/src/sm_c8y_mapper/http_proxy.rs

```diff
@@ -0,0 +1,350 @@
+            .bearer_auth(token.token())
+            .header("Accept", "application/json")
+            .timeout(Duration::from_millis(10000))
+            .build()?;
+
+        let response = client.execute(request).await?;
+        let event_response_body = response.json::<SmartRestLogEvent>().await?;
+
+        Ok(event_response_body.id)
+    }
+}
+
+#[async_trait]
+impl C8YHttpProxy for MqttAuthHttpProxy {
+    fn url_is_in_my_tenant_domain(&self, url: &str) -> bool {
+        self.end_point.url_is_in_my_tenant_domain(url)
+    }
+
+    #[instrument(skip(self), name = "init")]
+    async fn init(&mut self) -> Result<(), SMCumulocityMapperError> {
+        info!("Initialisation");
+        while self.end_point.c8y_internal_id.is_empty() {
+            if let Err(error) = self.try_get_and_set_internal_id().await {
+                error!(
+                    "An error ocurred while retrieving internal Id, operation will retry in {} seconds and mapper will reinitialise.\n Error: {:?}",
+                    RETRY_TIMEOUT_SECS, error
+                );
+
+                tokio::time::sleep(Duration::from_secs(RETRY_TIMEOUT_SECS)).await;
+                continue;
+            };
+        }
+        info!("Initialisation done.");
+        Ok(())
+    }
+
+    async fn get_jwt_token(&self) -> Result<SmartRestJwtResponse, SMCumulocityMapperError> {
+        let mut subscriber = self
+            .mqtt_con
+            .subscribe(Topic::new("c8y/s/dat")?.filter())
+            .await?;
+
+        let () = self
+            .mqtt_con
+            .publish(mqtt_client::Message::new(
+                &Topic::new("c8y/s/uat")?,
+                "".to_string(),
+            ))
+            .await?;
+
+        let token_smartrest =
+            match tokio::time::timeout(Duration::from_secs(10), subscriber.next()).await {
+                Ok(Some(msg)) => msg.payload_str()?.to_string(),
+                Ok(None) => return Err(SMCumulocityMapperError::InvalidMqttMessage),
+                Err(_elapsed) => return Err(SMCumulocityMapperError::RequestTimeout),
+            };
+
+        Ok(SmartRestJwtResponse::try_new(&token_smartrest)?)
+    }
+
+    async fn send_software_list_http(
+        &self,
+        c8y_software_list: &C8yUpdateSoftwareListResponse,
+    ) -> Result<(), SMCumulocityMapperError> {
+        let url = self.end_point.get_url_for_sw_list();
+        let client = reqwest::ClientBuilder::new().build()?;
+        let token = self.get_jwt_token().await?;
+
+        let request = client
+            .put(url)
+            .json(c8y_software_list)
+            .bearer_auth(&token.token())
+            .timeout(Duration::from_millis(10000))
+            .build()?;
+
+        let _response = client.execute(request).await?;
+
+        Ok(())
+    }
+
+    async fn upload_log_binary(
+        &self,
+        log_content: &str,
+    ) -> Result<String, SMCumulocityMapperError> {
+        let client = reqwest::ClientBuilder::new().build()?;
```

Comment:
It may not be part of this PR but it would be better if we reuse the client for the struct instead of creating it on each call to the method, `reqwest` client does internally manage resources, like connection pool etc anyway.
From their docs:
```plain
The Client holds a connection pool internally, so it is advised that you create one and reuse it.

You do not have to wrap the Client in an Rc or Arc to reuse it, because it already uses an Arc internally.
```

https://docs.rs/reqwest/0.11.7/reqwest/struct.Client.html

Reply:
That is a good point. And indeed something that can be done now we have a struct holding all the HTTP requests.